### PR TITLE
add docs for widgets

### LINKED ISSUE #155: AiiDAlab Applications: useful resources
Useful tools that can help you in building a powerful AiiDAlab application:

- AiiDAlab base widgets
- AiiDA plugins
- Ipywidgets
- Traitlets
...

### PR DESCRIPTION
**Pull request by dou-du**
fixes #155 

In this docs file, I write some information about the widgets:

* ipywidgets
* custom widgets
* aiidablab-widgets-base

### CODE REVIEW

File: docs/source/widgets/index.rst [hunk 1]
```diff
@@ -0,0 +1,105 @@
+###################
+Widgets
+###################
+
+The AiiDAlab largely dependents on the Jupyter widgets for the graphical user interface (GUI), namely the `Ipywidgets <https://ipywidgets.readthedocs.io/>`_ package.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:5]:
```suggestion
AiiDAlab largely dependents on Jupyter widgets for the graphical user interface (GUI).
More specifically, the `Ipywidgets <https://ipywidgets.readthedocs.io/>`_ package is the primary tool for creating AiiDAlab apps. 
```

File: docs/source/widgets/index.rst [hunk 2]
```diff
@@ -0,0 +1,105 @@
+###################
+Widgets
+###################
+
+The AiiDAlab largely dependents on the Jupyter widgets for the graphical user interface (GUI), namely the `Ipywidgets <https://ipywidgets.readthedocs.io/>`_ package. 
+Furthremore, we create the `aiidalab-widgets-base <https://github.com/aiidalab/aiidalab-widgets-base>`_ to facilitate to develop apps in AiiDAlab.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:6]:
```suggestion
We have also created the `aiidalab-widgets-base <https://github.com/aiidalab/aiidalab-widgets-base>`_ that is based on `Ipywidgets` and provides tools to interact with AiiDA.
```

File: docs/source/widgets/index.rst [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+***********************
+Ipywidgets
+***********************
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:10]:
```suggestion
**********
Ipywidgets
**********
```

File: docs/source/widgets/index.rst [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+`Ipywidgets <https://ipywidgets.readthedocs.io/>`_ is a python package to create widgets inside the Jupyter.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:12]:
```suggestion
`Ipywidgets <https://ipywidgets.readthedocs.io/>`_ is a python package that provides interactive widgets to be used in Jupyter notebooks. 
```

File: docs/source/widgets/index.rst [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+Furthremore, we create the `aiidalab-widgets-base <https://github.com/aiidalab/aiidalab-widgets-base>`_ to facilitate to develop apps in AiiDAlab.
+
+***********************
+Ipywidgets
+***********************
+
+`Ipywidgets <https://ipywidgets.readthedocs.io/>`_ is a python package to create widgets inside the Jupyter. 
+Widgets are eventful python objects, which shows GUI components such as a slider, button, etc.
+
+Ipywidgets provides base widgets like sliders, progressbar, buttons, checkbox, dropdown, radio buttons, text, textarea, label, etc.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:15]:
```suggestion
Ipywidgets package provides base widgets like sliders, progress bar, buttons, checkbox, dropdown, radio buttons, text, text area, label, etc. 
```

File: docs/source/widgets/index.rst [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+***********************
+Ipywidgets
+***********************
+
+`Ipywidgets <https://ipywidgets.readthedocs.io/>`_ is a python package to create widgets inside the Jupyter. 
+Widgets are eventful python objects, which shows GUI components such as a slider, button, etc.
+
+Ipywidgets provides base widgets like sliders, progressbar, buttons, checkbox, dropdown, radio buttons, text, textarea, label, etc. 
+The widgets have properties and value, which can be connected to the python codes and upated dynamically.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:16]:
```suggestion
The widgets have properties and value, which can be connected to the python codes and updated dynamically.
```

File: docs/source/widgets/index.rst [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    a.obersve(slider_change, names="value")
+
+In this example, the ``slider_change`` function will be called each time the slider' value being updated.
+
+Traitlets
+=================
+
+In ipywidgets, all the widgets are using the `traitlets <https://traitlets.readthedocs.io/>`_ types for the parameters.
+As shown in the example above, the value of the FloatSlider is a Float trait type.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:42]:
```suggestion
As shown in the example above, the value of the ``FloatSlider`` is a ``Float`` trait type.
```

File: docs/source/widgets/index.rst [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+One can not only show the current value ``value["new"]`` but also the old value ``value["old"]``.
+Traits can emit change events when the attributes are upated. This is very useful for creating custom widgets.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:44]:
```suggestion
Traits can emit change events when the attributes are updated.
This is very useful for creating custom widgets.
```

File: docs/source/widgets/index.rst [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+Custom widgets
+=================
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:47]:
```suggestion
Custom widgets
==============
```

File: docs/source/widgets/index.rst [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+Ipywidgets package also supports for developers to create their own widgets.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:49]:
```suggestion
Ipywidgets package also supports developers in creating their own widgets.
```

File: docs/source/widgets/index.rst [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+There is a detailed `tutorial <https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Custom.html>`_ to show how to develop the custom widgets.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:50]:
```suggestion
There is a detailed `tutorial <https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Custom.html>`_ to show how to develop a custom widget.
```

File: docs/source/widgets/index.rst [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+In the ipywidgets framework is based on the Comm framework.
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:51]: Please add a link.

File: docs/source/widgets/index.rst [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+Here is a list for some useful custom widgets:
+
+* `widget-periodictable <https://github.com/osscar-org/widget-periodictable>`_ : A interactive periodic table.
+* `widget-bandsplot <https://github.com/osscar-org/widget-bandsplot>`_ : A widget to plot bandstructure and density of states.
+* `widget-jsmol <https://github.com/osscar-org/widget-jsmol>`_ : A widget to use the molecular visualizer Jmol inside the Jupyter.
+
+***********************
+AiiDAlab base widgets
+***********************
```
Comment (yakutovicha) [inline docs/source/widgets/index.rst:92]:
```suggestion
*********************
AiiDAlab base widgets
*********************
```

**Review by yakutovicha — CHANGES_REQUESTED** [review]
Thanks @dou-du. I added some suggestions. Most of them are slight changes in the text.

To be honest, I am still not sure whether we want to put it as a separate section. Should it still be a part of the App development section?

@csadorf what do you think?

**Comment by csadorf** [comment]
> To be honest, I am still not sure whether we want to put it as a separate section. Should it still be a part of the App development section?
> 
> @csadorf what do you think?

Agreed, this should be a sub-section of the "App development" section.

File: docs/source/widgets/index.rst (hunk 12 above)
Comment (dou-du) [reply docs/source/widgets/index.rst:52]: The tutorial is hyperlink for the tutorial.

**Comment by yakutovicha** [comment]
@dou-du do you agree to move the section that you wrote under the "App development" section? If yes, can you do that?

**Comment by dou-du** [comment]
> @dou-du do you agree to move the section that you wrote under the "App development" section? If yes, can you do that?

Yes, I just move it to the "app development" section.

**Comment by yakutovicha** [comment]
thanks, @dou-du, just fix the pre-commit and we will merge it.

**Comment by yakutovicha** [comment]
@dou-du can you please fix the pre-commit issues?

**Review by yakutovicha — APPROVED** [review]
thanks a lot @dou-du !

**Pull request merged**